accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,781 @@
+
+The actual, underlying type of a modified type is obtained through the existing `Type.UnderlyingSystemType` property.
+
+This "modified type" terminology is used internally elsewhere including the CLR, `MetadataLoadContext` and `MetadataReader`. A "unmodified" type is a standard type that does not have any custom modifiers, even though in metadata it may.
+
+In CLI metadata, only fields, properties, local variables, method parameters and function pointer parameters allow custom modifiers. A type, method or function pointer itself cannot have custom modifiers; function pointers use the "return parameter" to contain its modopts (optional custom modifiers) used for calling conventions. Thus, the new `GetModified*Type()` methods could be called `GetFunctionPointerModified*Type()` instead. However, the shorter version is used here just in case this no longer holds to just function pointers.
+
+For additional information on encoding, see ECMA-335 `II.23.2 Blobs and signatures`. Note that in C# a function pointer is emitted as an ECMA "StandaloneMethodSig". There are several other types of custom modifiers besides the calling convention ones. For example, in C#, a field is also a "signature" and can have a `volatile` modifier which is emitted as a custom modifier and will be returned via `FieldInfo.GetModifiedFieldType().GetRequiredCustomModifiers()` as the `System.Runtime.CompilerServices.IsVolatile` type. In C++\CLI there are additional modifiers including `System.Runtime.CompilerServices.IsConst`.
+
+Supporting modified types correctly means also supporting "nested types" including
+- Arrays
+```cs
+    delegate*<int>[] a_0 = null!;
+    delegate*<int>[][] a_1 = null!;
+    delegate*<int>[,] a_2 = null!;
+```
+- Pointers
+```cs
+    delegate*<int>* p_0 = null;
+    delegate*<int>** p_1 = null;
+    delegate*<int>*** p_2 = null;
+```
+- Parameters of a parent function pointer
+```cs
+    delegate*<delegate*<int>> fp_0 = null;
+```
+- Combinations of the above
+```cs
+    // Arrays + pointers
+    delegate*<int>*[] ap_0 = null!;
+    delegate*<int>**[] ap_1 = null!;
+    delegate*<int>*[][] ap_2 = null!;
+
+    // Parent fcnptr + pointers
+    delegate*<delegate*<int>*> fp_1 = null;
+
+    // Parent fcnptr + arrays
+    delegate*<delegate*<int>[]> fp_2 = null;
+
+    // Parent fcnptr + arrays + pointers
+    delegate*<delegate*<int>*[]> fp_3 = null;
+```
+
+Note that a function pointer type cannot currently be used as a generic parameter type, so that limits this scope a bit.
```

Comment:
You can have arrays of function pointers as a generic parameter. For example, `new List<delegate*<int>[]>();` compiles fine. Is that going to work as expected?

Reply:
Yes.